### PR TITLE
newsolver alignment pipeline v2

### PR DESCRIPTION
**Pull request by trautmane**
I started refactoring the spark alignment pipeline components to simplify future development.  The first draft of the pipeline utilized a JSON configuration file in conjunction with a hard-coded ordering of components.  This draft introduces an AlignmentPipelineStep interface that clarifies what a pipeline step must support and allows step ordering to be specified in the JSON configuration file.  

If/when we are happy with this draft, the plan is to add intensity correction and thickness correction steps to the pipeline.

For review, I recommend starting by looking first at the [AlignmentPipelineStep interface](https://github.com/saalfeldlab/render/blob/ad30a5ed8d9488292da9d38c0bada88755b7f812/render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineStep.java) and how use of it simplifies the AlignmentPipelineClient.  There is also an [test/example JSON file](https://github.com/saalfeldlab/render/blob/ad30a5ed8d9488292da9d38c0bada88755b7f812/render-ws-spark-client/src/test/resources/pipeline/msem_alignment_pipeline.json) that might be useful.

### CODE REVIEW

File: render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineStep.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    /**
+     * Validates the specified pipeline parameters are sufficient for this step.
+     *
+     * @param  pipelineParameters  parameters to validate.
+     *
+     * @throws IllegalArgumentException
+     *   if any of the parameters are invalid or missing.
+     */
+    void validatePipelineParameters(final AlignmentPipelineParameters pipelineParameters)
+            throws IllegalArgumentException;
```
Comment (minnerbe) [inline render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineStep.java:23]: There are a few points that are not clear to me when I try to imagine how this method is used:
- Can it mutate the given `AlignmentPipelineParameters` or not? Either way, this should be made explicit, e.g. by simply stating it in the documentation, or with a `@Contract(pure = true)` annotation.
- The method does not return anything, so only the the "negative" outcome leads to observable behavior. Maybe returning a `bool` would be better? But then, `false` is redundant...

File: render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineStepId.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    GENERATE_MIPMAPS,
+    DERIVE_TILE_MATCHES,
+    PATCH_MFOV_MONTAGE_MATCHES,
+    FIND_UNCONNECTED_MFOVS,
+    FIND_UNCONNECTED_TILES_AND_EDGES,
+    FILTER_MATCHES,
+    ALIGN_TILES;
+
+    public AlignmentPipelineStep toStepClient()
+            throws UnsupportedOperationException {
```
Comment (minnerbe) [inline render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineStepId.java:27]: I usually like to give the enum a private field of type `Supplier<T>` instead of a huge switch statement. This requires a bit more effort up front, but makes the `toStepClient` method generic, such that adding a new enum doesn't have to be done in two places. E.g., define a new enum as `GENERATE_MIPMAPS(MipmapClient::new)` and have the `toStepClient` method only do
```Java
this.supplier.get()
```
Then, you can also get rid of the `UnsupportedOperationException`, since this case cannot happen anymore.

File: render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineParameters.java [hunk 3]
```diff
@@ -63,62 +73,48 @@ public MultiProjectParameters getMultiProject() {
         return multiProject;
     }
 
-    public boolean hasMipmapParameters() {
-        return mipmap != null;
-    }
-
```
Comment (minnerbe) [inline render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineParameters.java:69]: Why did you delete all those `has...` methods? Just asking to understand.
Edit: saw that these were only used in `AlignmentPipelineClient` to check if the pipeline step is to be executed. I like the current design much more!

File: render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineParameters.java [hunk 4]
```diff
@@ -63,62 +73,48 @@ public MultiProjectParameters getMultiProject() {
...
     public AffineBlockSolverSetup getAffineBlockSolverSetup() {
         return affineBlockSolverSetup;
     }
 
+    public List<AlignmentPipelineStep> buildStepClients()
+            throws IllegalArgumentException {
+
+        if ((pipelineSteps == null) || pipelineSteps.isEmpty()) {
+            throw new IllegalArgumentException("no pipeline steps defined");
+        }
```
Comment (minnerbe) [inline render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineParameters.java:109]: Just to make things clear: we go for "fail fast" rather than a smooth execution (in which case we could return an empty list), right? In this case, the check for missing parameters should probably be done right after JSON deserialization.
The exception should also be noted in the doc-string for this method.

File: render-ws-spark-client/src/test/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineParametersTest.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ */
+public class AlignmentPipelineParametersTest {
+
+    @Test
+    public void testJson()
+            throws IOException {
+        final AlignmentPipelineParameters pipelineParameters =
+                AlignmentPipelineParameters.fromJsonFile("src/test/resources/pipeline/msem_alignment_pipeline.json");
+
+        Assert.assertNotNull("deserialized parameters are null", pipelineParameters);
```
Comment (minnerbe) [inline render-ws-spark-client/src/test/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineParametersTest.java:23]: We should consider switching to JUnit5 in the long run: it always confuses me if I read something like "assertNotNull(".. are null")". JUnit5 places the `msg` parameter at the end of the call.

File: render-ws-spark-client/src/test/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineParametersTest.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+        Assert.assertEquals("incorrect affineBlockSolverSetup.targetStack.stackSuffix value parsed",
+                            "_align_pipe_alt_aa", affineBlockSolverSetup.targetStack.stackSuffix);
+
+        Assert.assertTrue("incorrect affineBlockSolverSetup.targetStack.completeTargetStack value parsed",
+                          affineBlockSolverSetup.targetStack.completeStack);
+
+        final List<AlignmentPipelineStep> stepClients = pipelineParameters.buildStepClients();
+        Assert.assertNotNull("stepClients is null", stepClients);
+        Assert.assertEquals("incorrect number of stepClients", 7, stepClients.size());
```
Comment (minnerbe) [inline render-ws-spark-client/src/test/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineParametersTest.java:36]: This test method does a lot of things:
- checking if JSON deserialization produces a non-null object;
- verify that the correct target stack settings are loaded;
- testing the conversion to `AlignmentPipelineStep`.
Should we maybe split this into several tests with more descriptive names? This way the test could also serve as kind of documentation.

File: render-ws-spark-client/src/main/java/org/janelia/render/client/spark/match/ClusterCountClient.java [hunk 7]
```diff
@@ -46,63 +48,80 @@ public org.janelia.render.client.ClusterCountClient buildJavaClient() {
...
+    public void validatePipelineParameters(final AlignmentPipelineParameters pipelineParameters)
+            throws IllegalArgumentException {
+        AlignmentPipelineParameters.validateRequiredElementExists("tileCluster",
+                                                                  pipelineParameters.getTileCluster());
     }
 
-    public void run() throws IOException {
+    /** Run the client as part of an alignment pipeline. */
+    public void runPipelineStep(final JavaSparkContext sparkContext,
+                                final AlignmentPipelineParameters pipelineParameters)
```
Comment (minnerbe) [inline render-ws-spark-client/src/main/java/org/janelia/render/client/spark/match/ClusterCountClient.java:90]: I really like how the changed handling of the parameter object now makes the steps seem like proper pipeline steps (the parameters now act as shared context instead of internal state of the steps).

File: render-ws-spark-client/src/main/java/org/janelia/render/client/spark/match/CopyMatchClient.java [hunk 8]
```diff
@@ -47,56 +49,63 @@ public Parameters(final MultiProjectParameters multiProject,
...
 
-    public void copyMatches(final JavaSparkContext sparkContext)
-            throws IOException {
+    /** Validates the specified pipeline parameters are sufficient. */
+    @Override
+    public void validatePipelineParameters(final AlignmentPipelineParameters pipelineParameters)
+            throws IllegalArgumentException {
+        AlignmentPipelineParameters.validateRequiredElementExists("matchCopy",
+                                                                  pipelineParameters.getMatchCopy());
+    }
```
Comment (minnerbe) [inline render-ws-spark-client/src/main/java/org/janelia/render/client/spark/match/CopyMatchClient.java:88]: The validation seems to only verify that the parameters are present, not that they are valid. Is the name maybe too general for the purpose it serves?

File: render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineParameters.java [hunk 9]
```diff
@@ -50,6 +59,7 @@ public AlignmentPipelineParameters(final MultiProjectParameters multiProject,
                                        final MatchCopyParameters matchCopy,
                                        final AffineBlockSolverSetup affineBlockSolverSetup) {
```
Comment (minnerbe) [inline render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineParameters.java:60]: This hardcodes all existing steps, isn't it? Is there a way around that? I guess every approach that deals with those in a more generic (and dynamic) way probably loses type information, which would be a strong argument against doing that.

File: render-ws-spark-client/src/main/java/org/janelia/render/client/spark/multisem/MFOVMontageMatchPatchClient.java [hunk 10]
```diff
@@ -43,84 +44,86 @@ public static class Parameters extends CommandLineParameters {
...
             @Override
             public void runClient(final String[] args) throws Exception {
-
                 final Parameters parameters = new Parameters();
                 parameters.parse(args);
-
-                final MFOVMontageMatchPatchClient client = new MFOVMontageMatchPatchClient(parameters);
-                client.run();
+                final MFOVMontageMatchPatchClient client = new MFOVMontageMatchPatchClient();
+                client.createContextAndRun(parameters);
```
Comment (minnerbe) [inline render-ws-spark-client/src/main/java/org/janelia/render/client/spark/multisem/MFOVMontageMatchPatchClient.java:57]: Does every client create its own context? If so, do we want that (I'm not familiar enough with Spark contexts to answer this question myself)?

File: render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineStep.java (hunk 1 above)
Comment (trautmane) [reply render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineStep.java:26]: I tried to add some more-helpful javadocs.  Bottom line is no mutation and throw exception for problem.  This is a validation pattern/approach that makes sense to me and is repeated in many other contexts.

File: render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineStepId.java (hunk 2 above)
Comment (trautmane) [reply render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineStepId.java]: Great idea - done.

File: render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineParameters.java (hunk 4 above)
Comment (trautmane) [reply render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineParameters.java:115]: Yes - fail fast.  This check is effectively done right after deserialization.  Added javadoc - good point.

File: render-ws-spark-client/src/test/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineParametersTest.java (hunk 6 above)
Comment (trautmane) [reply render-ws-spark-client/src/test/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineParametersTest.java:38]: Yeah - I was being lazy.  I've split things up now.

File: render-ws-spark-client/src/main/java/org/janelia/render/client/spark/match/CopyMatchClient.java (hunk 8 above)
Comment (trautmane) [reply render-ws-spark-client/src/main/java/org/janelia/render/client/spark/match/CopyMatchClient.java:88]: My intent/hope is that we flesh-out the validation implementations to do more later - assuming that is useful/helpful.

File: render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineParameters.java (hunk 9 above)
Comment (trautmane) [reply render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineParameters.java:60]: Yeah - I was going for explicit declaration in a small number of places rather than a pure generic solution.

File: render-ws-spark-client/src/main/java/org/janelia/render/client/spark/multisem/MFOVMontageMatchPatchClient.java (hunk 10 above)
Comment (trautmane) [reply render-ws-spark-client/src/main/java/org/janelia/render/client/spark/multisem/MFOVMontageMatchPatchClient.java:57]: For backwards compatibility, I was trying to keep support for running each client on its own outside of the new pipeline.  This is why there are createContextAndRun methods.  However, you are correct that when run as a pipeline, a single spark context will be shared and passed to the runPipelineStep method.

Comment (minnerbe) [reply render-ws-spark-client/src/main/java/org/janelia/render/client/spark/multisem/MFOVMontageMatchPatchClient.java:57]: Thanks for the clarification! You are right, being able to run the clients on their own is desirable.

File: render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineStep.java (hunk 1 above)
Comment (minnerbe) [reply render-ws-spark-client/src/main/java/org/janelia/render/client/spark/pipeline/AlignmentPipelineStep.java:26]: I wasn't familiar with that pattern, so I tried to come up with a "better" solution for several minutes, but I didn't succeed. So, I feel convinced that this should stay as it is. Your javadoc does help a lot, though!

**Comment by minnerbe** [comment]
I had a look at the latest commits (and your comments), they look very good to me. From my side, I think we can go ahead and merge this.
